ksonnet-gen/ksonnet/component.go

```diff
@@ -0,0 +1,66 @@
+package ksonnet
+
+import (
+	"github.com/go-openapi/spec"
+	"github.com/pkg/errors"
+)
+
+const (
+	extensionGroupVersionKind = "x-kubernetes-group-version-kind"
+)
+
+// Component is resource information provided in the k8s swagger schema
+// which contains the group, kind, and version for a definition.
+type Component struct {
+	Group   string
+	Kind    string
+	Version string
+}
+
+// NewComponent extracts component information from a schema.
+func NewComponent(s spec.Schema) *Component {
+	re := componentExtractor{schema: s}
+	group := re.extract("group")
+	kind := re.extract("kind")
+	version := re.extract("version")
+
+	if re.err != nil {
+		return nil
+	}
+
+	return &Component{
+		Group:   group,
+		Kind:    kind,
+		Version: version,
+	}
+}
+
+type componentExtractor struct {
+	err    error
+	schema spec.Schema
+}
+
+func (re *componentExtractor) extract(key string) string {
+	if re.err != nil {
+		return ""
+	}
+
+	i, ok := re.schema.Extensions[extensionGroupVersionKind]
+	if !ok {
+		re.err = errors.New("no group/kind/version extension")
+		return ""
+	}
+
+	s, ok := i.([]interface{})
+	if ok {
+		m, ok := s[0].(map[string]interface{})
+		if ok {
```

Comment:
if `!ok` for these, do we need to set an error?

Reply:
I was torn about this. This was part of a bigger piece of code, so in the next round or two, it might be changed.